File: frontend/src/api/courses.ts
/**
 * Course endpoint wrappers.
 *
 * Three read-only header lookups used by the profile wizard + Discovery
 * filter bar. All routes are course-scoped under `/courses/{id}/*` and
 * already RLS-respecting on the backend.
 */

import { apiFetch } from "@/api/client";
import type { CourseSummary } from "@/types/api";

export interface SectionRead {
  id: string;
  code: string;
}

export interface CourseSkillRead {
  id: string;
  skill_name: string;
  display_order: number;
}

export function getCourse(courseId: string): Promise<CourseSummary> {
  return apiFetch<CourseSummary>(`/courses/${courseId}`);
}

export function listSections(courseId: string): Promise<SectionRead[]> {
  return apiFetch<SectionRead[]>(`/courses/${courseId}/sections`);
}

export function listCourseSkills(courseId: string): Promise<CourseSkillRead[]> {
  return apiFetch<CourseSkillRead[]>(`/courses/${courseId}/skills`);
}

File: backend/app/api/v1/courses.py
"""Course metadata routes.

Sibling to ``discovery.py`` — both share the ``/courses`` prefix.
Discovery owns the People + Groups feeds; this module owns the three
header lookups (course detail, sections list, skill catalog) the
profile wizard and Discovery filter bar both consume.

All routes use ``user_session`` so RLS automatically scopes results to
"courses I'm enrolled in". 404 covers both "doesn't exist" and
"filtered by RLS" so we don't leak the existence of other courses.
"""

from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, HTTPException, status

from app.db.session import CurrentUserDep, UserSessionDep
from app.schemas.courses import CourseSkillRead, CourseSummary, SectionRead
from app.services import courses as courses_service

router = APIRouter(prefix="/courses", tags=["courses"])


def _err(status_code: int, code: str, message: str) -> HTTPException:
    return HTTPException(status_code=status_code, detail={"code": code, "message": message})


@router.get(
    "/{course_id}",
    response_model=CourseSummary,
    summary="Course header (code, name, deadline, timezone)",
)
async def get_course(
    course_id: UUID,
    db: UserSessionDep,
    _user: CurrentUserDep,
) -> CourseSummary:
    try:
        return await courses_service.get_course(db, course_id)
    except courses_service.CourseNotFound as exc:
        raise _err(
            status.HTTP_404_NOT_FOUND,
            "COURSE_NOT_FOUND",
            "Course not found or not visible to you.",
        ) from exc


@router.get(
    "/{course_id}/sections",
    response_model=list[SectionRead],
    summary="List sections in the course",
)
async def list_sections(
    course_id: UUID,
    db: UserSessionDep,
    _user: CurrentUserDep,
) -> list[SectionRead]:
    try:
        return await courses_service.list_sections(db, course_id)
    except courses_service.CourseNotFound as exc:
        raise _err(
            status.HTTP_404_NOT_FOUND,
            "COURSE_NOT_FOUND",
            "Course not found or not visible to you.",
        ) from exc


@router.get(
    "/{course_id}/skills",
    response_model=list[CourseSkillRead],
    summary="Course skill catalog (used by profile wizard + Discovery filter)",
)
async def list_skills(
    course_id: UUID,
    db: UserSessionDep,
    _user: CurrentUserDep,
) -> list[CourseSkillRead]:
    try:
        return await courses_service.list_skills(db, course_id)
    except courses_service.CourseNotFound as exc:
        raise _err(
            status.HTTP_404_NOT_FOUND,
            "COURSE_NOT_FOUND",
            "Course not found or not visible to you.",
        ) from exc
